x-pack/plugins/siem/public/components/url_state/index.tsx

```diff
@@ -0,0 +1,439 @@
+      getQueryStringFromLocation(newLocation),
+      'urlStateKey'
+    );
+
+    if (previousUrlStateString !== newUrlStateString) {
+      const previousUrlState = mapToUrlState(decodeRisonUrlState(previousUrlStateString));
+      const newUrlState = mapToUrlState(decodeRisonUrlState(newUrlStateString));
+
+      if (typeof newUrlState !== 'undefined') {
+        onChange(newUrlState, previousUrlState);
+      }
+    }
+  };
+}
+
+export const UrlStateComponents = pure<UrlStateContainerProps>(props => (
+  <Route<RouteProps>>
+    {({ history, location }) => (
+      <UrlStateContainerLifecycle
+        data-test-subj={'urlStateComponents'}
+        history={history}
+        location={location}
+        {...props}
+      />
+    )}
+  </Route>
+));
+
+const makeMapStateToProps = () => {
+  const getInputsSelector = inputsSelectors.inputsSelector();
+  const getHostsFilterQueryAsKuery = hostsSelectors.hostsFilterQueryAsKuery();
+  const getNetworkFilterQueryAsKuery = networkSelectors.networkFilterQueryAsKuery();
+  const mapStateToProps = (state: State) => {
+    const inputState = getInputsSelector(state);
+    const { linkTo: globalLinkTo, timerange: globalTimerange } = inputState.global;
+    const { linkTo: timelineLinkTo, timerange: timelineTimerange } = inputState.timeline;
+
+    return {
+      urlState: {
+        timerange: {
+          global: {
+            timerange: globalTimerange,
+            linkTo: globalLinkTo,
+          },
+          timeline: {
+            timerange: timelineLinkTo,
+            linkTo: timelineTimerange,
+          },
+        },
+        kqlQuery: [
+          {
+            filterQuery: getHostsFilterQueryAsKuery(state, hostsModel.HostsType.details),
+            type: hostsModel.HostsType.details,
+            model: KueryFilterModel.hosts,
+          },
+          {
+            filterQuery: getHostsFilterQueryAsKuery(state, hostsModel.HostsType.page),
+            type: hostsModel.HostsType.page,
+            model: KueryFilterModel.hosts,
+          },
+          {
+            filterQuery: getNetworkFilterQueryAsKuery(state, networkModel.NetworkType.details),
+            type: networkModel.NetworkType.details,
+            model: KueryFilterModel.network,
+          },
+          {
+            filterQuery: getNetworkFilterQueryAsKuery(state, networkModel.NetworkType.page),
+            type: networkModel.NetworkType.page,
+            model: KueryFilterModel.network,
+          },
+        ],
+      },
+    };
+  };
+
+  return mapStateToProps;
+};
+export const UrlStateContainer = connect(
+  makeMapStateToProps,
+  {
+    setAbsoluteTimerange: inputsActions.setAbsoluteRangeDatePicker,
+    setHostsKql: hostsActions.applyHostsFilterQuery,
+    setNetworkKql: networkActions.applyNetworkFilterQuery,
+    setRelativeTimerange: inputsActions.setRelativeRangeDatePicker,
+    toggleTimelineLinkTo: inputsActions.toggleTimelineLinkTo,
+  }
+  // @ts-ignore
+)(UrlStateComponents);
```

Comment:
There are 2 remaining ts issues, this is one of them. I've spent the better part of the past 6 hours trying to figure this out. May need to merge and come back. Both @spong and @XavierM have spent time pairing on this, it's a toughy:
<img width="1100" alt="Screen Shot 2019-05-21 at 10 11 18 AM" src="https://user-images.githubusercontent.com/6935300/58121210-15240780-7bc4-11e9-858b-c0e5d9cd958d.png">